src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestExpressions.ll

```diff
@@ -0,0 +1,72 @@
+  %18 = call i1 @Microsoft__Quantum__Testing__QIR__ReturnEquality__body(i64 4, i64 5)
+  %19 = call i1 @Microsoft__Quantum__Testing__QIR__ReturnInequality__body(i64 5, i64 6)
+  %20 = call i1 @Microsoft__Quantum__Testing__QIR__ReturnLessThan__body(i64 5, i64 6)
+  %21 = call i1 @Microsoft__Quantum__Testing__QIR__ReturnLessThanOrEqual__body(i64 5, i64 6)
+  %22 = call i1 @Microsoft__Quantum__Testing__QIR__ReturnGreaterThan__body(i64 5, i64 6)
+  %23 = call i1 @Microsoft__Quantum__Testing__QIR__ReturnGreaterThanOrEqual__body(i64 5, i64 6)
+  %24 = call i1 @Microsoft__Quantum__Testing__QIR__ReturnLogicalAnd__body(i1 true, i1 false)
+  %25 = call i1 @Microsoft__Quantum__Testing__QIR__ReturnLogicalOr__body(i1 false, i1 true)
+  %26 = call double @Microsoft__Quantum__Testing__QIR__ReturnAddition__body(double 1.000000e+00, double 2.000000e+00)
+  %27 = call i64 @Microsoft__Quantum__Testing__QIR__ReturnSubtraction__body(i64 3, i64 4)
+  %28 = call double @Microsoft__Quantum__Testing__QIR__ReturnMultiplication__body(double 1.000000e+00, double 2.000000e+00)
+  %29 = call i64 @Microsoft__Quantum__Testing__QIR__ReturnDivision__body(i64 3, i64 4)
+  %30 = call i64 @Microsoft__Quantum__Testing__QIR__ReturnExponentiation1__body(i64 5)
+  %31 = call double @Microsoft__Quantum__Testing__QIR__ReturnExponentiation2__body(double 6.000000e+00)
+  %32 = call %BigInt* @Microsoft__Quantum__Testing__QIR__ReturnExponentiation3__body(i64 7)
+  %33 = call i64 @Microsoft__Quantum__Testing__QIR__ReturnModulo__body(i64 8)
+  %34 = call i2 @Microsoft__Quantum__Testing__QIR__ReturnPauli__body()
+  %35 = call %Result* @Microsoft__Quantum__Testing__QIR__ReturnResult__body()
+  %36 = call i1 @Microsoft__Quantum__Testing__QIR__ReturnBool__body()
+  %37 = call i64 @Microsoft__Quantum__Testing__QIR__ReturnInt__body()
+  %38 = call double @Microsoft__Quantum__Testing__QIR__ReturnDouble__body()
+  %39 = call %BigInt* @Microsoft__Quantum__Testing__QIR__ReturnBigInt__body()
+  call void @Microsoft__Quantum__Testing__QIR__ReturnUnit__body()
+  %40 = call i64 @Microsoft__Quantum__Testing__QIR__ReturnLeftShift__body()
+  %41 = call %BigInt* @Microsoft__Quantum__Testing__QIR__ReturnRightShift__body()
+  %42 = call i64 @Microsoft__Quantum__Testing__QIR__ReturnBXOr__body()
+  %43 = call i64 @Microsoft__Quantum__Testing__QIR__ReturnBOr__body()
+  %44 = call %BigInt* @Microsoft__Quantum__Testing__QIR__ReturnBAnd__body()
+  %45 = call i1 @Microsoft__Quantum__Testing__QIR__ReturnNot__body()
+  %46 = call i64 @Microsoft__Quantum__Testing__QIR__ReturnBNot__body()
+  %47 = call double @Microsoft__Quantum__Testing__QIR__ReturnNegative__body(double 3.000000e+00)
+  call void @__quantum__rt__callable_unreference(%Callable* %0)
+  call void @__quantum__rt__callable_unreference(%Callable* %1)
+  call void @__quantum__rt__callable_unreference(%Callable* %4)
+  %48 = bitcast { %TupleHeader, { %TupleHeader, i2, i64 }*, double }* %5 to %TupleHeader*
```

Comment:
when is the inner TupleHeader released?

Reply:
Looking more closely, I believe the idea was that a call to tuple_unreference, callable_unreference and similar would recur and also call the unreference for inner tuples and capture tuples respectively. However, I don't think the runtime actually can do that, since I believe it may not have sufficient type information to do so. Hence, I think we may need to actually handle that in the QIR emission. @alan-geller Am I missing something?

Reply:
For the tuples, the runtime doesn't know the types and cannot unreference inner elements. For the callables, the runtime could unreference the capture tuple, but it should be called out in the QIR spec: 1. whether the runtime should increment the reference of the given tuple, when creating a callable, and 2. that it should decrement the reference when releasing the callable.

Reply:
Note that some of this is covered in issue #641 .

Reply:
Array items also need to be unreferenced and can't be by the runtime since it won't know the their type.